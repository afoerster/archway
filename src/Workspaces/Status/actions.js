export const APPROVE_WORKSPACE_REQUESTED = 'APPROVE_WORKSPACE_REQUESTED';

export function approveInfra() {
  return {
    type: APPROVE_WORKSPACE_REQUESTED,
    role: 'infra',
  };
}

export function approveRisk() {
  return {
    type: APPROVE_WORKSPACE_REQUESTED,
    role: 'risk',
  };
}

export const APPROVE_WORKSPACE_COMPLETED = 'APPOVE_WORKSPACE_COMPLETED';

export function approveWorkspaceCompleted(error) {
  return {
    type: APPROVE_WORKSPACE_COMPLETED,
    error,
  };
}
