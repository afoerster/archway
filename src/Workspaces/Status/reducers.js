import {
  APPROVE_WORKSPACE_COMPLETED,
  APPROVE_WORKSPACE_REQUESTED,
} from './actions.js';

const initialState = {
  approving: false,
};

const status = (state = initialState, action) => {
  switch (action.type) {
    case APPROVE_WORKSPACE_REQUESTED:
      return {
        ...state,
        approving: true,
      };
    case APPROVE_WORKSPACE_COMPLETED:
      return {
        ...state,
        approving: false,
      };
    default:
      return state;
  }
}

export default status;
