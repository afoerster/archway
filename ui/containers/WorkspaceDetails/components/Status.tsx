import * as React from 'react';
import { Icon } from 'antd';
import { Colors } from '../../../components';

/* tslint:disable:no-var-requires */
const TimeAgo = require('timeago-react').default;

interface Props {
  ready?: boolean;
  createdAt: Date;
}

const Status = ({ ready, createdAt }: Props) => (
  <div style={{ display: 'flex', alignItems: 'center' }}>
    <Icon
      style={{ fontSize: 32 }}
      type={ready ? 'check-circle' : 'close-circle'}
      theme="twoTone"
    />
    <div style={{ textTransform: 'uppercase', marginLeft: 8 }}>
      <div style={{ fontSize: 18 }}>
        {ready ? 'READY' : 'PENDING'}
      </div>
      <div style={{ fontSize: 10, color: Colors.Gray.string(), textTransform: 'uppercase' }}>
        created <TimeAgo datetime={createdAt} />
      </div>
    </div>
  </div>
);

export default Status;
