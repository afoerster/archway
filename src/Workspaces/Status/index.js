import React from 'react';
import { connect } from 'react-redux';
import {
  Row,
  Col,
  Icon,
  Button,
} from 'antd';

import Panel from '../Panel';
import { approveRisk, approveInfra } from './actions';

const ComplianceCheck = ({ label, present }) => (
  <div style={{ display: 'flex', flexDirection: 'column', justifyItems: 'space-between' }}>
    <Icon style={{ fontSize: 28, color: present ? "#FF5900" : "#0B7A75" }} type={present ? "warning" : "minus-circle-o"} />
    <h3>{label}</h3>
  </div>
);

const ApprovalActions = ({ approving, approved, canApprove, approve, label }) => (
  <Panel style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}>
    <div style={{
      fontWeight: 100,
      fontSize: 24,
      color: approved ? "#0B7A75" : "#FF5900",
      overflow: 'hidden'
    }}>
      {approved ? "approved" : "pending"}
    </div>
    <div>
      {label}
    </div>
    {canApprove && (
      <Button
        icon="check"
        disabled={approved}
        loading={approving}
        type="primary"
        size="large"
        onClick={approve}>
        Approve {label}
      </Button>
    )}
  </Panel>
);

const Status = ({
  approving,
  approveInfra,
  approveRisk,
  workspace: {
    status,
    approvals = {},
    compliance: {
      pii_data,
      phi_data,
      pci_data,
    }
  },
  profile: {
    permissions: {
      risk_management,
      platform_operations,
    }
  }
}) => (
  <div>
    <h2>Approvals</h2>
    <hr />
    <Row type="flex" justify="center">
      <Col span={8}>
        <ApprovalActions
          label="Infrastructure"
          approving={approving}
          approved={approvals && approvals.infra}
          canApprove={platform_operations}
          approve={approveInfra} />
      </Col>
      <Col span={8}>
        <ApprovalActions
          label="Risk"
          approving={approving}
          approved={approvals && approvals.risk}
          canApprove={risk_management}
          approve={approveRisk} />
      </Col>
    </Row>
    <h2>Compliance</h2>
    <hr />
    <Row type="flex" justify="center">
      <Col span={4}>
        <ComplianceCheck
          label="PII"
          present={pii_data} />
      </Col>
      <Col span={4}>
        <ComplianceCheck
          label="PCI"
          present={pci_data} />
      </Col>
      <Col span={4}>
        <ComplianceCheck
          label="PHI"
          present={phi_data} />
      </Col>
    </Row>
  </div>
);

export default connect(
  state => state.workspaces.status, { approveRisk, approveInfra }
)(Status);
