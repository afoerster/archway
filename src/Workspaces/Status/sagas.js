import { call, select, fork, takeLatest, put } from 'redux-saga/effects';

import * as Api from '../../API';
import {
  APPROVE_WORKSPACE_REQUESTED,

  approveWorkspaceCompleted
} from './actions';

import { getWorkspace } from '../WorkspaceDetails/actions';

function* requestApproval({ role }) {
  const token = yield select(s => s.auth.token);
  const { id } = yield select(s => s.workspaces.details.activeWorkspace);
  let error;
  try {
    yield call(Api.approveWorkspace, token, id, role);
  } catch (e) {
    error = e;
  }
  yield put(approveWorkspaceCompleted(error));
  yield put(getWorkspace(id));
}

function* approveWorkspace() {
  yield takeLatest(APPROVE_WORKSPACE_REQUESTED, requestApproval);
}

export default function* root() {
  yield fork(approveWorkspace);
}