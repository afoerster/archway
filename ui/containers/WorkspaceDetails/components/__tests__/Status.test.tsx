import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Status from '../Status';

describe('Status', () => {
  it('renders correctly', () => {
    const props = {
      ready: true,
      createdAt: new Date('2019-06-20T20:52:47Z'),
    };
    const wrapper = shallow(<Status {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
