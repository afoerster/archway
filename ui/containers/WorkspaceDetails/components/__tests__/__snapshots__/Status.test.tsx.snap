// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status renders correctly 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
    }
  }
>
  <Icon
    style={
      Object {
        "fontSize": 32,
      }
    }
    theme="twoTone"
    type="check-circle"
  />
  <div
    style={
      Object {
        "marginLeft": 8,
        "textTransform": "uppercase",
      }
    }
  >
    <div
      style={
        Object {
          "fontSize": 18,
        }
      }
    >
      READY
    </div>
    <div
      style={
        Object {
          "color": "rgb(187, 187, 187)",
          "fontSize": 10,
          "textTransform": "uppercase",
        }
      }
    >
      created 
      <TimeAgo
        datetime={2019-06-20T20:52:47.000Z}
        live={true}
        locale="en"
      />
    </div>
  </div>
</div>
`;
